R/epi_shift_internal.R

```diff
@@ -0,0 +1,218 @@
+#'  be the lag or lead for each value in the vector. Lag integers must be
+#'  nonnegative, while ahead integers must be positive.
+#' @param prefix A prefix to indicate what type of variable this is
+#' @param default Determines what fills empty rows
+#'   left by leading/lagging (defaults to NA).
+#' @param keys A character vector of the keys in an epi_df
+#' @param columns A character string of variable names that will
+#'  be populated (eventually) by the `terms` argument.
+#' @param skip A logical. Should the step be skipped when the
+#'  recipe is baked by [bake()]? While all operations are baked
+#'  when [prep()] is run, some operations may not be able to be
+#'  conducted on new data (e.g. processing the outcome variable(s)).
+#'  Care should be taken when using `skip = TRUE` as it may affect
+#'  the computations for subsequent operations.
+#' @param id A unique identifier for the step
+#' @template step-return
+#'
+#' @details The step assumes that the data are already _in the proper sequential
+#'  order_ for shifting.
+#'
+#' The `prefix` and `id` arguments are unchangeable to ensure that the code runs
+#'  properly and to avoid inconsistency with naming. For `step_epi_ahead`, they
+#'  are always set to `"ahead_"` and `"epi_ahead"` respectively, while for
+#'  `step_epi_lag`, they are set to `"lag_"` and `"epi_lag`, respectively.
+#'
+#' @family row operation steps
+#' @rdname step_epi_shift
+#' @export
+#' @examples
+#' epi_recipe(case_death_rate_subset) %>%
+#'   step_epi_ahead(death_rate, ahead = 7) %>%
+#'   step_epi_lag(death_rate, lag = c(0,7,14))
+step_epi_lag <-
+  function(recipe,
+           ...,
+           role = "predictor",
+           trained = FALSE,
+           lag = 1,
+           prefix = "lag_",
+           default = NA,
+           keys = epi_keys(recipe),
+           columns = NULL,
+           skip = FALSE,
+           id = rand_id("epi_lag")) {
+    stopifnot("Lag values must be nonnegative integers" =
+                all(lag>=0 & lag == as.integer(lag)))
+
+    step_epi_shift(recipe,
+                   ...,
+                   role = role,
+                   trained = trained,
+                   shift = lag,
+                   prefix = prefix,
+                   default = default,
+                   keys = keys,
+                   columns = columns,
+                   skip = skip,
+                   id = id
+    )
+  }
+
+#' Create a shifted predictor
+#'
+#' @family row operation steps
+#' @rdname step_epi_shift
+#' @export
+step_epi_ahead <-
+  function(recipe,
+           ...,
+           role = "outcome",
+           trained = FALSE,
+           ahead = 1,
+           prefix = "ahead_",
+           default = NA,
+           keys = epi_keys(recipe),
+           columns = NULL,
+           skip = FALSE,
+           id = rand_id("epi_ahead")) {
+
+    stopifnot("Ahead values must be positive integers" =
+                all(ahead>0 & ahead == as.integer(ahead)))
```

Comment:
`ahead == 0` should be ok.

Reply:
I intentionally did `ahead > 0` as that'd be the only way to unambiguously distinguish a `lag` from an `ahead` given that it's possible to change the names of the parameters, but that the shift value distinguishes between the two.

How should I go about it? I tried doing many different things, but this should be fail-safe while satisfying guidelines.